tests/Equinox.MessageDb.Integration/MessageDbIntegration.fs

```diff
@@ -0,0 +1,481 @@
+        let service2 = Cart.createServiceWithSnapshottingAndCaching log context cache
+
+        // Trigger 8 events, then reload
+        let cartId = % Guid.NewGuid()
+        do! addAndThenRemoveItemsManyTimes ctx cartId skuId service1 4
+        let! _ = service2.Read cartId
+
+        // ... should not see a snapshot write as we are inside the batch threshold
+        listener.TestSpans([|
+            span ["name", "Transact"; "eqx.batches", 1; "eqx.count", 0; "eqx.snapshot_version", -1L
+                  "eqx.cache_hit", null; "eqx.append_count", 8; "eqx.snapshot_written", null]
+            span ["name", "Query"; "eqx.batches", 1; "eqx.count", 8; "eqx.snapshot_version", -1L
+                  "eqx.cache_hit", false]
+
+        |])
+
+        // Add two more, which should push it over the threshold and hence trigger generation of a snapshot event
+        do! addAndThenRemoveItemsManyTimes ctx cartId skuId service1 1
+        listener.TestSpans([|
+            span ["name", "Transact"; "eqx.batches", 1; "eqx.count", 8; "eqx.snapshot_version", -1L
+                  "eqx.cache_hit", null; "eqx.append_count", 2; "eqx.snapshot_written", true]
+        |])
+        // We now have 10 events, we should be able to read them with a single snapshotted read
+        let! _ = service1.Read cartId
+        listener.TestSpans([|
+            span ["name", "Query"; "eqx.batches", 1; "eqx.count", 1; "eqx.snapshot_version", 10L
+                  "eqx.cache_hit", null]
+        |])
+
+        // Add 8 more; total of 18 should not trigger snapshotting as the snapshot is at version 10
+        do! addAndThenRemoveItemsManyTimes ctx cartId skuId service1 4
+        listener.TestSpans([|
+            span ["name", "Transact"; "eqx.batches", 1; "eqx.count", 1; "eqx.snapshot_version", 10L
+                  "eqx.cache_hit", null; "eqx.append_count", 8; "eqx.snapshot_written", null]
+        |])
+
+        // While we now have 18 events, we should be able to read them with a single snapshotted read
+        let! _ = service1.Read cartId
+        listener.TestSpans([|
+            span ["name", "Query"; "eqx.batches", 1; "eqx.count", 9; "eqx.snapshot_version", 10L
+                  "eqx.cache_hit", null]
+        |])
+
+        // ... trigger a second snapshotting
+        do! addAndThenRemoveItemsManyTimes ctx cartId skuId service1 1
+        // and we _could_ reload the 20 events with a single read. However we are using the cache, which last saw it with 10 events, which necessitates two reads
+        let! _ = service2.Read cartId
+        listener.TestSpans([|
+            span ["name", "Transact"; "eqx.batches", 1; "eqx.count", 9; "eqx.snapshot_version", 10L
+                  "eqx.cache_hit", null; "eqx.append_count", 2; "eqx.snapshot_written", true]
+            span ["name", "Query"; "eqx.batches", 2; "eqx.count", 12; "eqx.snapshot_version", null
+                  "eqx.cache_hit", true]
+        |])
+    }
+
```

Comment:
these tests do look clean - static/async local state FTW!
except for these blank lines!
Also clean all Rider warnings please (Serilog.Log vs Log etc - maybe have one alias for `let log = Serilog.Core.NullLogger` or whatever at top level)